### ISSUE TITLE
Git: Support WSL

### ISSUE THREAD

**Issue by tianvan** [Microsoft/vscode#9502]
![qq 20160720033734](https://cloud.githubusercontent.com/assets/10600430/16963929/70f48324-4e2b-11e6-9d03-e683978d7989.png)

**Comment by joaomoreno** [comment]
This was fast.

**Comment by jbw91** [comment]
It would be great to use Bash on Ubuntu on Windows as the integrated terminal! Not to mention its git installation for the git tab. I can't +1 this enough! :)

**Comment by jbw91** [comment]
Noticed that we can use the Bash on Ubuntu on Windows terminal in the integrated terminal now. That's awesome! Now we just need the git integration with the git version installed through bash on ubuntu on windows.

**Comment by isaacDaly** [comment]
+1

**Comment by ncarroll** [comment]
+1

**Comment by nmbradford** [comment]
+1

**Comment by jonomacd** [comment]
This would be great not just for git but for extensions as well.  I use the vscode-go extension and it shells out to use the command line tools often.  I'd like it if it could shell out to bash as opposed to windows terminal.  It would also need access to environment variables from bash.  Basically, I want it to think it is in a bash environment even though it is in windows.  

Thanks.

**Comment by ElieMyIdea** [comment]
+1

**Comment by ElieMyIdea** [comment]
+1

**Comment by zhcm** [comment]
+1

**Comment by andre2** [comment]
+1

**Comment by doxiaodong** [comment]
+1

**Comment by lobabob** [comment]
+1

**Comment by EmmanuelVlad** [comment]
+1

**Comment by kidpollo** [comment]
+1

**Comment by mikhailbalin** [comment]
+1

**Comment by lugolven** [comment]
+1 :)

**Comment by sgiradoa-psl** [comment]
+1

**Comment by skippednote** [comment]
@joaomoreno Wanted to check if there any work being done on this WSL git + VSCode integration?

**Comment by nickjj** [comment]
With Insider's 1.19 from Nov 15th I was able to track git commits with having VSCode installed on Windows while having git installed on WSL and do basically everything except pull/push (due to VSCode not using SSH keys from WSL) without any configuration or hacks.

But, with today's Nov 20th update now it's complaining about the `git.path` not finding a suitable version of git. However, it only checks Windows for a git installation instead of WSL.

Since this feature was working a few days ago, what should the new `git.path` be set to in order to integrate with WSL's git? i tried all sorts of bash.exe related paths without success.

**Comment by CoenraadS** [comment]
@nickjj Just set it to your Windows git.exe. VSCode never had support for WSL git. You can still use most of the features, but it will use the windows config not the WSL config.

**Comment by nickjj** [comment]
@CoenraadS I never had Git installed on Windows (ever).

But prior to this patch, the editor was capable of showing changes in a repo where git was only installed on WSL.

**Comment by CoenraadS** [comment]
Prior to the latest insiders it checked the PATH, so if you do cmd -> `where git.exe` you might be surprised.

**Comment by nickjj** [comment]
`where git.exe` returns nothing from PowerShell, because I don't have Git installed on Windows.

If I run `which git` from WSL it's found at `/usr/bin/git`.

**Comment by CoenraadS** [comment]
I discovered this great project: https://github.com/andy-5/wslgit for people who want this support today.

**Comment by problem233** [comment]
+1

**Comment by erend** [comment]
Integrating this feature make web coding on Windows perfect.

**Comment by OmriSama** [comment]
+1

**Comment by seashore001x** [comment]
+1

**Comment by jbw91** [comment]
Hey everyone, friendly reminder to use the thumbs up reaction to the original comment instead of everyone posting +1. :) That's how interest is gauged, not by the number of comments with +1 which is harder to quantify. (Also doesn't spam everyone's notifications) ;)

**Comment by nikhilweee** [comment]
So with the new year, is there a native solution yet? Or is wslgit still the best workaround for now?

**Comment by mikerockett** [comment]
Looks like this hasn’t made it to the [roadmap](/Microsoft/vscode/wiki/Roadmap). :cry:

**Comment by Stopa** [comment]
+1

**Comment by trickymast3r** [comment]
Im workin in this !

**Comment by erikmueller** [comment]
Just a naïve approach: I tried installing Git for windows, then added the `Program Files/Git/bin` to my path and created an alias for `git.exe=git`. 
Since WSL can run windows binaries both (`wsl.exe` and `vscode`) now use the windows git installation. 
Maybe this helps and sorry if I got the problem wrong ^^

**Comment by pd93** [comment]
@erikmueller I believe this issue is for people who are using the WSL as their development environment and don't want to have to install Git for Windows in addition to WSL Git.

In other words, VSCode should be able to use WSL Git as it's SCM rather than requiring users to install additional software (Git for Windows).

It's worth noting that the integrated terminal can already use WSL Git, but that the SCM cannot.

**Comment by trickymast3r** [comment]
I'm done with this but ... VSCode so hard to commit my code ... sad !

**Comment by pd93** [comment]
@trickymast3r That's great news. Have you got a fork/MR everyone could take a look at?

**Comment by jedmao** [comment]
@trickymast3r I had a similar thing going on in Node.js. I used [pkg](https://github.com/zeit/pkg) to package-up my `git.js` file into a `git.exe` file that ran a `git.cmd` file with the following contents:

```bat
@echo off
wsl git %*
@echo on
```

Vscode isn't complaining that Git doesn't exist now, but it *is* complaining that there are "no active source control providers."

I also tried `mklink git.exe git.cmd`. No 🎲🎲  !

**Comment by trusktr** [comment]
Can someone please take this out of Backlog? It's fairly relevant. Especially if you want to entice Linux/macOS devs to use Windows...

**Comment by Cifram** [comment]
Yes, was just struggling with this and am surprised to find it still not addressed. I'd really not have to maintain my ssh keys in two separate places with two installs of git. VS Code is great about working with WSL in so many ways, but this is still missing.

**Comment by hungryphilomath** [comment]
+1

**Comment by wasabigeek** [comment]
@Cifram I've gotten away with one set of SSH keys by creating one in `/mnt/` and using it in both WSL and Windows - maybe that's an option to explore? This doesn't remove the need to install Git on both WSL and Windows though.

**Comment by MrBlenny** [comment]
I'm doing the same thing.
I gave [wslgit](https://github.com/andy-5/wslgit) but it was far too slow to be usable.

Instead installed GIT on windows as well as WSL.
Ssh keys are stored in wsl.
A symlink/junction was created in windows to link to wsl. Your path will vary.
```cmd
mklink /J C:\Users\{User}\.ssh C:\Users\{User}\AppData\Local\Packages\CanonicalGroupLimited.UbuntuonWindows_79rhkp1fndgsc\LocalState\rootfs\home\{user}\.ssh
```

I have heard that editing linux files from windows can cause problems so changes to these ssh keys should probably only be made from WSL.

**Comment by sudesh-poojari** [comment]
+1

**Comment by patch153** [comment]
+1

**Comment by iksemot** [comment]
I've been looking into this and, I think, I made a little bit of progress. Maybe someone will be able to take this further.

The way it should work is to set `"git.path": "wsl git"` in User Settings of VSC as this is the preferred way of running WSL commands [1]. However there seem to be a problem with node child_process when spawning a command with spaces (needs confirmation).

The quite obvious solution, as already suggested [2], is to create a cmd script:
```
@echo on
wsl git %*
@echo off
```
and then use it in User Settings;  `"git.path": "C:\\wsl\\git.cmd"`.

However there is some problem with that. Turns out that spawning a process which starts wsl, for some reason makes child_process to wait for an input. And that prevents VSC to finish the process of locating git executable.

I monkey-patched `resources\app\extensions\git\out\git.js` function `findSpecificGit` to ignore stdin
```
options = {stdio: ['ignore', null, null]};
const child = cp.spawn(path, ['--version'], options);
```
And that made VSC to successfully recognize git! But now there's an issue with recognizing git repository.

What VSC does, it goes through various directories, starting with project directory and executes command `wsl git rev-parse --show-toplevel` which returns an absolute path to the git repository. But it returns it in Unix-lie format. And here I'm stuck.

[1] https://docs.microsoft.com/en-us/windows/wsl/interop#run-linux-tools-from-a-windows-command-line
[2] https://github.com/Microsoft/vscode/issues/9502#issuecomment-364043243

**Comment by pd93** [comment]
@iksemot Not sure if this is what you're looking for, but since Windows 10 [17046](https://docs.microsoft.com/en-us/windows/wsl/release-notes#build-17046) there is a tool called `wslpath` which will convert between windows/wsl paths. This will be included in the upcoming (April) release of Windows 10 (RS4).

This binary resides in the wsl bin, so perhaps something like this might work?
`wsl wslpath -wa $(git rev-parse --show-toplevel)`
That will convert the value returned by git to an absolute windows path.

**Comment by nikhilweee** [comment]
On the issue of `child_process` waiting for an input, does https://github.com/nodejs/node/issues/18431 seem to help?

**Comment by Audiosyncrasy** [comment]
👍

**Comment by kabala** [comment]
👍

**Comment by victor-perez** [comment]
what you can do is first try `wslpath` if that fails you try a normal `git` command. Is not ideal but it works
see [Use WSL git inside VS Code from Windows 10 17046
](https://gist.github.com/victor-perez/8ba25540394ec68b11f4b2699fb4d4cel)

**Comment by mterrel** [comment]
@victor-perez Your solution worked great for me. Thanks!

**Comment by schankam** [comment]
2 years later, still no reliable option to do this simply with "git.path". Quite annoying...

**Comment by bardware** [comment]
Do you start `code` from `wsl` in the first place or do you start code from `cmd`?

**Comment by Lo1icon** [comment]
sshkey please

**Comment by mroach** [comment]
What could be nice is an option to run all or select external binaries with WSL. For example if you're using code linters or runners they'll require things like Node (+ package bins), Ruby (+ gem bins), Elixir (+ package bins), PostgreSQL, etc. Currently the only way to get these working is installing them natively in Windows which is totally redundant and may not always produce the same results.

Currently the only practical way I can see around this is running a Linux VM with VS Code inside it.

**Comment by ameft** [comment]
@mroach there's a specific open issue about that: #22663. Let's add some pressure.. ;P

**Comment by tprasadtp** [comment]
I messed with some lines of Go and I could make it work with vs code. Its ugly hack and a lot of edge cases are not considered as well. Its [here](https://github.com/tprasadtp/wsl-git/releases), for some reason vs code does not like buffered output, it looks ugly. Its not perfect but does the job for me anyway.

**Comment by problem233** [comment]
@tprasadtp Here is already a workaround: https://github.com/andy-5/wslgit

**Comment by hangxingliu** [comment]
I wrote a solution batch scripts project: <https://github.com/hangxingliu/wslgit> 

It is inspired by a Rust project [andy-5/wslgit](https://github.com/andy-5/wslgit). And you can use it by copy and paste without any compiler.

**Comment by grtrout** [comment]
@hangxingliu What benefit does your project have over andy-5's? Have you tested it to be faster?

**Comment by khuongduybui** [comment]
Andy's project was created before wslpath was published by wsl itself. Most of the logic was about path conversion. Now that wslpath is a provided too, we don't really need to perform much logic or computation at all, so rust provides no benefits of significance. Meanwhile bash can be used anywhere without compiling or installing any additional dependencies. I think that's the main benefit.

**Comment by pd93** [comment]
@khuongduybui It's worth noting that the BASH solution does require `gawk` so technically does have a dependency.

Also, the Rust implementation does not require you to have the compile tools. You can simply place the `exe` downloaded from the releases page into your Windows PATH somewhere. The BASH solution actually requires you to place two files in two different locations; one in the WSL and one in your Windows PATH.

**Comment by khuongduybui** [comment]
I wouldn't say it's *better*. I was simply explaining why it's *different*. Whether a difference is considered better, worse, or irrelevant is entirely up to the environment it's applied in, so I won't argue with you.

**Comment by pd93** [comment]
@khuongduybui Agreed. and your point about performance still stands. I just thought I'd add some additional info.

**Comment by pd93** [comment]
@grtrout I've been using separate installations for Windows and the WSL for around 6 months or so now and the only real issues I've come across are keeping the `.gitconfig` files and SSH keys synced.

I also have `core.autocrlf` set to `input` in both configs as this ensures that line endings are consistent no matter which version of Git I use.

While Andy's project works well for me, I still feel more comfortable using two official tools over a single unofficial one until VSC adds proper support. This is definitely down to the individual though.

**Comment by khuongduybui** [comment]
@grtrout I don't think wslpath is required for any of these "wslgit" solutions.

**Comment by grtrout** [comment]
@pd93 Great info - I will make that `core.autocrlf` change to prevent future headache. Regarding keeping SSH keys synced - without thinking much about it, I created one in WSL in addition the Windows one I already had. I added both to my GitHub/GitLab profiles. Is there any reason it would make more sense to try syncing the same key across environments? If I misunderstood, let me know. Thanks.

@khuongduybui Thanks for the clarification. I made the assumption that wslpath was needed based on your comment:  
> Now that wslpath is a provided tool, we don't really need to perform much logic or computation at all, so rust provides no benefits of significance.

**Comment by haddigan** [comment]
@grtrout I have been using Andy's solution and although it is much slower than having a separate git installation, that hasn't been a non-starter for me, just a minor inconvenience. (I have tested @khuongduybui 's solution as well and performing actions takes just about as long).

As @pd93 mentioned maintaining separate ssh keys and config across two different installations of git seems to be the main issue for most people. This is basically a situation where all the tools to make this happen are available, and it seems like a no-brainer as far as including it in an update. MS has provided all of these great tools to make development on Windows much better, but it just feels incomplete as long as those tools can't interoperate with one another.

**Comment by khuongduybui** [comment]
@grtrout I'm sorry for the confusion. I checked the code; none of them used wslpath.
@haddigan it's not MY solution. Creator is @hangxingliu

**Comment by hangxingliu** [comment]
Sorry for reply late, because I am on a business trip.

The reason why I created repository <https://github.com/hangxingliu/wslgit> (reference from <https://github.com/andy-5/wslgit>) are:

1. I am used to the Linux environment for develop, but I bought a LG gram laptop and it is hard to installed Ubuntu with ACPI. so I decided use Windows and WSL.
2. I don't want to maintain two sets of git config and SSH config between Windows and WSL. (and also problems about CRLF and Linux file permissions)
3. I have not learned Rust development. So It is hard to setup Rust environment and compile Andy's wslgit on Windows for me. (And Windows SmartScreen prevent me execute its pre-compiled binary file)
4. I chose bash to implement this repository because I am familiar with it, and you can use it very easy without compile. (And I will remove dependency  `gawk` from repository because it can working well with `awk`  and also `gawk`)

And I saw many people are discussing use `wslpath` instead of programmatic path conversion(using `awk` or `Rust`). But I think `wslpath` is hard to find which part of git output is the path that needs to be converted. And It is hard to decide which parameter for git is a path. (And `wslpath` will failed if its input parameter is not a valid path).

I also found a bug when converting path doesn't implement by `wslpath`: you changed mount point of Windows drive in `/etc/wsl.conf`. And I am thinking about how to resolve this problem.   
And if anyone has good idea about this, I am very welcome discuss with you.

@grtrout @pd93 @khuongduybui 

---
P.s. My most desirable solution is VSCode supports git in WSL.

**Comment by pd93** [comment]
@hangxingliu `wslpath` works absolutely fine with a custom mount point in the `wsl.conf` file. That's one of the advantages of using it. I've been using it like this for a while now.

Also, you can easily test if a command was successful or not by checking the error code. `wslpath` will return a non zero code if it fails. eg:
```shell
wslpath /bin
echo $?
# Output: 1
```

**Comment by sudesh-poojari** [comment]
@pd93 Can you share your setup or link on how to setup git using `wslpath`

**Comment by pd93** [comment]
@sudesh-poojari I should clarify. I am not currently using Git _with_ `wslpath`. I'm still using separate Git installations for Windows and WSL. However, I do sometimes use `wslpath` for other things.

There is an [open issue](https://github.com/andy-5/wslgit/issues/12) for implementing `wslpath` into [Andy's project](https://github.com/andy-5/wslgit), but there are a couple of blocking problems that need to be worked out. Take a look at the issue comments for details.

**Comment by hangxingliu** [comment]
@pd93 Thanks for your guide. I will push a new commit in my repository late. include fixed a bug about Windows batch file (escaped characters) and `wslpath` support for input parameters.

By the way, Do you have a good idea about convert linux path in git output to Windows path by using `wslpath`?

**Comment by hangxingliu** [comment]
@pd93 That's right. Sometimes the output of git is only a path string, for example:

``` bash
> git rev-parse --show-toplevel
/mnt/c/repo1
```

But there are some conditions: the output of git is not only a path string, also many other description text .and maybe includes more than one path, for example:

``` bash
> git remote -v
local   /mnt/c/local_remote (fetch)
local   /mnt/c/local_remote (push)
origin  git@github.com:hangxingliu/wslgit.git (fetch)
origin  git@github.com:hangxingliu/wslgit.git (push)
```

So It is hard to extract which part is a Linux path be mounted from Windows drive by using `wslpath` only  in the second condition(The output included more than one path or included other strings)

**Comment by hangxingliu** [comment]
@pd93 Up to latest commit in my [wslgit](https://github.com/hangxingliu/wslgit) repository. It supported convert Linux path in git output to Windows path by `wslpath` or `awk` (replace path following WSL mounted `drvfs` list).

**Comment by andy-5** [comment]
For anyone having performance issues with [wslgit](https://github.com/andy-5/wslgit) or similar tools, the slowdown is most likely caused by the execution of your Bash startup scripts (like `.bashrc`).

For [andy-s/wslgit](https://github.com/andy-5/wslgit) specifically, you can force it to not start an interactive Bash shell, which skips executing `.bashrc`, by setting the Windows environment variable `WSLGIT_USE_INTERACTIVE_SHELL` to `false` (as documented [here](https://github.com/andy-5/wslgit#advanced-usage)). This can significantly speed up `wslgit`, but does not run any tools like `ssh-agent`, etc. you may have added to your `.bashrc`.

**Comment by mickeypuri** [comment]
Discussing with a colleague, he has simply set in vscode user preferences, the terminal.integrated.shell.windows to point to the bash.exe for WSL, and then set git.path in vscode preferences to the path in the WSL where git is installed e.g. 
`    
"terminal.integrated.shell.windows": "C:\\Windows\\System32\\bash.exe",
 "git.path": "C:\\Users\\some_user_name\\AppData\\Local\\lxss\rootfs\\usr\\bin\\git",
`
It appears to work however he is not using the vscode git plugin as it felt flaky. Is this a potential solution or is something being missed in this setup (as I know we are not meant to alter any files in the linux system using any windows tools)?

(One point on the git.path, more recent installs of bash, will use a different path rather than the lxss folder [as outlined here](https://github.com/Microsoft/WSL/issues/2578) )

**Comment by galcianuk** [comment]
Would also like support for this!

**Comment by thomaswolff** [comment]
Support for using Git installation in WSL would be great!

**Comment by duttaditya18** [comment]
Can anyone take a look at this Gist by @victor-perez : https://gist.github.com/victor-perez/8ba25540394ec68b11f4b2699fb4d4ce

It doesn't work for me. Can anyone else test it out?

**Comment by duttaditya18** [comment]
Figured it out by using this : https://github.com/andy-5/wslgit

**Comment by mike-zipit** [comment]
This is working for me:  https://gist.github.com/victor-perez/8ba25540394ec68b11f4b2699fb4d4ce

**Comment by phifa** [comment]
@joaomoreno Integrating WSL GIT into VS Code would be amazing! Any updates here?

**Comment by mmirus** [comment]
@phifa I've been using it in VS Code for quite a while. IIRC all you need to do is save the latest `wslgit.exe` release to your computer and then tell VS Code to use it in your settings:

```json
"git.path": "C:\\tools\\wslgit.exe",
```

**Comment by murillopaula** [comment]
@mmirus Yes, but we want to use it without downloading and anything other than WSL itself.

**Comment by nickjj** [comment]
Keep in mind if you go with the `wslgit.exe` approach, it works but it's deadly slow. So slow that it's completely unusable (even when things outside of Git was very fast).

**Comment by phifa** [comment]
@joaomoreno is a native integration in VSC planned?

**Comment by iamruhul** [comment]
# It's 2019.. It's been about 3 years since this issue was brought up... We see no progress at all.
 Git with vscode is the only reason I am in linux.. using wsl and git for windows together causes some crazy problems like losing permissions on some folders/files. So please vscode guys do something.

**Comment by mmirus** [comment]
I know it's not a permanent solution but if you haven't tried wslgit with vscode, give it a shot. It's stable and performs quite well for me.

**Comment by phifa** [comment]
Well I think a heads up would be good. Is it being worked on? Can we contribute or test? Don’t leave us in the dark 🙂

**Comment by pd93** [comment]
@phifa I think @mmirus is referring to https://github.com/andy-5/wslgit which is an unofficial workaround

**Comment by benc-uk** [comment]
wslgit is slow as hell and it's a cludgy workaround. 

Also SCM/git is just one part of the WSL integration we need from VSCode

**Comment by ChristianGrimberg** [comment]
> Figured it out by using this : https://github.com/andy-5/wslgit
Works great for me :+1:

WSL & Git & Visual Studio Code without Git Bash

**Comment by phifa** [comment]
@joaomoreno any updates on this?

**Comment by llatseattle** [comment]
Any update on this?

**Comment by andreialecu** [comment]
[wslgit](https://github.com/andy-5/wslgit) is not slow at all in my experience if set up properly.

Just remember to set the `WSLGIT_USE_INTERACTIVE_SHELL` variable in Windows to `false`, see the documentation in its README.

**Comment by panoscool** [comment]
the integrated terminal is working fine with the wsl the issue is the vscode git sidebar is not working at all with the wsl

**Comment by quhung1989** [comment]
@pan This is what I did to resolve the side bar issue. I opened "Git Bash" in Windows and and ran:

git config --global core.excludesfile '/the/absolute/path/to/your/global/git/ignore'

So currently Git Bash and WSL git config --global are referencing the same ignore file. I was not sure how Git Bash handles absolute paths so I did "cd ~" and "$PWD" and it returned "/c/Users/username".

**Comment by RMacfarlane** [comment]
We just announced remote development with VS Code, check out the blog post for details https://code.visualstudio.com/blogs/2019/05/02/remote-development

**Comment by xtydev** [comment]
> We just announced remote development with VS Code, check out the blog post for details https://code.visualstudio.com/blogs/2019/05/02/remote-development

@RMacfarlane  Thank you for the heads-up. It really looks great! I have one question though:
Now with the **[Visual Studio Code Remote - WSL extension](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-wsl)** in action, will VS Code automatically pick up the Git installation in WSL or do we have to manually configure `"git.path"` in the [WSL specific settings](https://code.visualstudio.com/docs/remote/wsl#_wsl-specific-settings)?

**Comment by benc-uk** [comment]
It automatically finds git from the WSL system path, no need to configure anything

**Comment by 12101111** [comment]
With remote-wsl extension, vscode will use toolchains in wsl, but sometimes we just want to use git in wsl and continue to use the toolchain in Windows.
![图片](https://user-images.githubusercontent.com/8438475/57138198-a9305980-6de4-11e9-9a6a-effccebc83f9.png)

**Comment by andreialecu** [comment]
One probably helpful piece of information regarding the new `Remote - WSL` extension, which I needed to do in my project:

If you open your folder from the windows path, like you used to - such as from your recent projects menu, the extension won't kick in - everything will still be in Windows mode.

What I had to do was to use `Remote-WSL: Open Window`, then the `File > Open Folder...` menu will show you Linux paths, so you need to browse for your folder from there (`/mnt/...`) and open that instead.

After doing this the git used by Code will be the native one from WSL, and your terminal windows will also use bash by default.

**Comment by panoscool** [comment]
@andreialecu if you open the project from wsl terminal using code . wsl Git is working right?

**Comment by andreialecu** [comment]
@panoscool `code-insiders .` doesn't seem to work properly from the WSL terminal. It still opens via the Windows path, meaning the extension doesn't kick in - so you don't get WSL Git or any other WSL capabilities.

It also prints an error to the console, which `code .` doesn't:

```
 $ code-insiders . 
/c/Users/<snip>/AppData/Local/Programs/Microsoft VS Code Insiders/bin/code-insiders: line 29: cmd.exe: command not found
```

**Comment by andreialecu** [comment]
### My experience after a few days with Remote-WSL (valid May 2019):

I went back to using `wslgit` on stable Code. While Remote-WSL is extremely promising, I kept running into file system issues, such as various EACCES errors, and even hard lockups because of the NTFS<->WSL file access issues and VS Code was exacerbating them as well.

One example (using React Native) was weird lockups of the Android emulator in Android Studio that could only be resolved by a computer reboot. Additionally, Android Studio would rarely build the project without complaining about being unable to replace some file because of access denied issues. Had to randomly close/reopen Code to try to get past these errors.

One other annoying thing for example is that you are unable to rename a directory after opening it via Remote-WSL. It's a known limitation, among others listed here: https://code.visualstudio.com/docs/remote/wsl#_common-limitations-in-wsl

**These will all likely be resolved once WSL 2.0, which was just announced, goes live. It should hit the Windows Insider channel in late June: https://devblogs.microsoft.com/commandline/announcing-wsl-2/**

In the mean time, `wslgit`, while it has its quirks, is currently less of a hindrance to work with.

**Comment by aeschli** [comment]
Thanks for the feedback, @andreialecu. Yes, the fs locking issues are nasty. As you said, they will only go away with the VM2 mode. In some cases you can work around the issue by defining watcher excludes ('files.watcherExclude") for folders where other tools work (e.g. node_modules...)

For the cmd issue you created https://github.com/microsoft/vscode-remote-release/issues/170, thanks!
